whylog/constraints/verifier.py

```diff
@@ -0,0 +1,110 @@
+from whylog.constraints import IdenticalIntervals, TimeConstraint
+from whylog.exceptions import WhylogError
+from whylog.front import FrontInput
+
+
+class ConstraintManager(object):
+    @classmethod
+    def get_constraint_by_type(cls, constraint_data):
+        constraints = {
+            'identical': IdenticalIntervals({}),
+            'time': TimeConstraint({})
+            # register your constraint here
+        }
+        if constraint_data['name'] in constraints:
+            return constraints[constraint_data['name']]
+        raise WhylogError("No such constraint (%s) registered" % constraint_data['name'])
+
+
+class Verifier(object):
+    @classmethod
+    def _front_input_from_clue(cls, clue):
+        return FrontInput(clue.line_offset, clue.line_prefix_content, clue.line_source)
+
+    @classmethod
+    def _verify_constraint(cls, combination, effect, constraint):
+        constraint_verifier = ConstraintManager.get_constraint_by_type(constraint)
+        groups = []
+        for group_info in constraint['clues_groups']:
+            parser_num, group_num = group_info[0], group_info[1]
+            if parser_num == 0:
+                groups.append(effect.regex_parameters[group_num - 1])
+            else:
+                groups.append(combination[parser_num - 1].regex_parameters[group_num - 1])
+        return constraint_verifier.verify(constraint['params'], groups)
+
+    @classmethod
+    def _clues_combinations(cls, clues_lists, collected_subset=[]):
+        """
+        recursive generator that returns all combinations
+        of elements from lists contained in clues_lists
+        example:
+        >>> xs = [[1, 2, 3], [4, 5], [6]]
+        >>> for l in Verifier._clues_combinations(xs):
+        >>>     print l
+        [1, 4, 6]
+        [1, 5, 6]
+        [2, 4, 6]
+        [2, 5, 6]
+        [3, 4, 6]
+        [3, 5, 6]
+        it always should be called with empty accumulator,
+        that is collected_subset=[]
+        """
+        if len(clues_lists) != 0:
+            for clue in clues_lists[0]:
+                for subset in cls._clues_combinations(clues_lists[1:], collected_subset + [clue]):
+                    yield subset
+        else:
+            yield collected_subset
+
+    @classmethod
+    def constraints_and(cls, clues_lists, effect, constraints):
+        clues_lists = list(filter(lambda x: x, clues_lists))
```

Comment:
This line causes bug. clues_lists has unformal mapping clues_lists[i] gets Clues which come from i position in rule causes parsers. When some parsers was unmatched. In clues you got empty list on its index. This filtering line destroy this mapping. I think that combinations function must get non empty list, what cause this filtering.